name: build
on: [push, pull_request, workflow_dispatch]

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        tests: ['', 'external']
        dependencies: ['dev', 'deploy']
    steps:
    # - name: Cancel Previous Runs
    #   uses: styfle/cancel-workflow-action@0.5.0
    #   with:
    #     access_token: ${{ github.token }}

    - uses: actions/checkout@v2

    - name: Setup node
      uses: actions/setup-node@v1
      with:
        node-version: '12'

    - name: Docker compose up
      run: docker-compose up -d

    - name: Set up Python 3.6
      uses: actions/setup-python@v2
      with:
        python-version: 3.6

    - name: Install Poetry
      uses: snok/install-poetry@v1.1.1
      with:
        version: 1.0.10

    # - name: Use Cache
    #   uses: actions/cache@v2
    #   with:
    #     path: |
    #       ~/.cache/pip
    #       ~/.cache/pypoetry/virtualenvs
    #       ~/.npm
    #     key: ${{ runner.os }}-pip-venv-${{ hashFiles('**/poetry.lock') }}

    - name: Bootstrap
      if: ${{ matrix.dependencies == 'dev' }}
      run: |
        poetry run bootstrap --ci
        poetry install --no-root --extras sip2
        poetry run pip install --upgrade coveralls

    - name: Bootstrap deploy
      if: ${{ matrix.dependencies == 'deploy' }}
      run: |
        poetry run bootstrap --ci --deploy E2E=yes
        poetry install --no-root --extras sip2
        poetry run pip install --upgrade coveralls

    - name: Run Test
      if: ${{ matrix.tests == '' }}
      run: poetry run run-tests

    - name: Run Test external
      if: ${{ matrix.tests == 'external'}}
      id: test_external
      continue-on-error: true
      run: poetry run run-tests external

    # NOT WORKING BECAUSE OF THIS PROBLEM:
    # https://github.community/t/github-actions-are-severely-limited-on-prs/18179#M9249
    # - uses: LouisBrunner/checks-action@v1.1.0
    #   if: ${{ steps.test_external.outcome == 'failure' }}
    #   id: status
    #   with:
    #     token: ${{ secrets.GITHUB_TOKEN }}
    #     name: Set test external failure
    #     conclusion: neutral
    #     output: |
    #       { "summary": "TEST EXTERNAL FAILED!!!" }

    - name: Upload Coverage ${{ matrix.tests }}
      if: ${{ matrix.dependencies == 'dev' }}
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        COVERALLS_SERVICE_NAME: github
        COVERALLS_FLAG_NAME: test-${{ matrix.tests }}
        COVERALLS_PARALLEL: true
      run: poetry run coveralls


  coveralls:
    needs: [test]
    runs-on: ubuntu-latest
    steps:
    - name: Cancel Previous Runs
      uses: styfle/cancel-workflow-action@0.5.0
      with:
        access_token: ${{ github.token }}

    - name: Set up Python 3.6
      uses: actions/setup-python@v2
      with:
        python-version: 3.6

    - name: Finished
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        COVERALLS_SERVICE_NAME: github
        COVERALLS_PARALLEL_FINISHED: true
      run: |
        pip install --upgrade coveralls
        coveralls --finish
